resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -100,11 +97,47 @@ protected DnsNameResolverContext(DnsNameResolver parent,
         allowedQueries = maxAllowedQueries;
     }
 
-    protected Promise<T> promise() {
-        return promise;
+    void resolve(Promise<T> promise) {
+        boolean directSearch = parent.searchDomains().length == 0 || StringUtil.endsWith(hostname, '.');
+        if (directSearch) {
+            internalResolve(promise);
+        } else {
+            final Promise<T> original = promise;
+            promise = parent.executor().newPromise();
+            promise.addListener(new FutureListener<T>() {
+                int count;
+                @Override
+                public void operationComplete(Future<T> future) throws Exception {
+                    if (future.isSuccess()) {
+                        original.trySuccess(future.getNow());
+                    } else {
+                        if (count < parent.searchDomains().length) {
+                            String searchDomain = parent.searchDomains()[count++];
+                            Promise<T> nextPromise = parent.executor().newPromise();
+                            String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
+                            DnsNameResolverContext<T> nextContext = newResolverContext(parent,
+                                nextHostname, resolveCache);
+                            nextContext.internalResolve(nextPromise);
+                            nextPromise.addListener(this);
+                        } else {
+                            original.tryFailure(future.cause());
+                        }
+                    }
+                }
+            });
+            int idx = hostname.length();
+            int dots = 0;
+            while (idx-- > 0) {
```

Comment:
any reason why you don't use a `for` loop here? Indexing seems a bit "unconventional".

```java
for (int i = hostname.length() - 1; i >= 0; --i) {

}
```